test/test_torch.py

```diff
@@ -6977,15 +6977,64 @@ def compare_strides(s1, s2, div):
 
     @skipMeta
     @dtypes(*torch.testing.get_all_dtypes())
-    def test_dlpack_conversion(self, device, dtype):
-        # DLpack does not explicitly support bool
+    def test_dlpack_capsule_conversion(self, device, dtype):
+        # DLpack does not explicitly support bool (xref dmlc/dlpack#75)
         # It does it through uint8 type
-        if dtype is torch.bool:
+        if dtype is torch.bool or 'xla' in device:
             return
         x = make_tensor((5,), device, dtype, low=-9, high=9)
         z = from_dlpack(to_dlpack(x))
         self.assertEqual(z, x)
 
+    @skipMeta
+    @dtypes(*torch.testing.get_all_dtypes())
+    def test_dlpack_protocol_conversion(self, device, dtype):
+        if dtype is torch.bool or 'xla' in device:
+            return
+        x = make_tensor((5,), device, dtype, low=-9, high=9)
+        z = from_dlpack(x)
+        self.assertEqual(z, x)
+
+    @skipMeta
+    @dtypes(*torch.testing.get_all_dtypes())
+    def test_dlpack_conversion_with_streams(self, device, dtype):
+        if dtype is torch.bool or 'xla' in device:
+            return
+        # Create a stream where the tensor will reside
+        if device == 'cuda':
+            x = make_tensor((5,), device, dtype, low=-9, high=9)
+            stream = torch.cuda.Stream()
+            with torch.cuda.stream(stream):
+                z = from_dlpack(x)
+            stream.synchronize()
+            self.assertEqual(z, x)
+
+    @skipMeta
+    @dtypes(*torch.testing.get_all_dtypes())
+    def test_dlpack_conversion_with_diff_streams(self, device, dtype):
+        if dtype is torch.bool or 'xla' in device:
+            return
+        if device == 'cuda':
+            from torch._C import _from_dlpack
+            x = make_tensor((5,), device, dtype, low=-9, high=9)
+            stream_a = torch.cuda.Stream()
+            stream_b = torch.cuda.Stream()
+            with torch.cuda.stream(stream_a):
+                z = _from_dlpack(x.__dlpack__(stream_b.cuda_stream))
```

Comment:
What are the semantics here? x is converted to a dlpack tensor on stream b, and then z is constructed from the dlpack tensor on stream a? In that case, stream a needs to wait for stream b to complete is work

Reply:
here, `__dlpack__` will add an event to `stream_b` to enforce ordering with `stream_a`

https://github.com/pytorch/pytorch/blob/4e90170386a80278a25e7027a742db66ebf7f035/torch/_tensor.py#L1055-L1060